### ISSUE TITLE
Limit parallelism of sstable probing

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#1880]
column_family::populate(sstring sstdir) does this:

```c++
        return lister::scan_dir(sstdir, { directory_entry_type::regular }, [this, sstdir, verifier, descriptor, &futures] (directory_entry de) {
             auto f = probe_file(sstdir, de.name).then([verifier, descriptor, sstdir, de] (auto entry) {
             ...
            futures.push_back(std::move(f));
```

So all the sstables are probed in parallel. There could be thousands of sstables, and since probing requires opening a file, this can cause thousands of files (in each shard) to be open at the same time. Not terrible, but not ideal either.

I think we should use a semaphore (as described in the Seastar tutorial ;-)) to limit the parallelism of this process - there's really no point of probing more than, say, 100 files in parallel.

Moreover, note that database::populate_keyspace() calls this column_family::populate() for all the CFs in parallel. So the limit to column_family::populate()'s parallelism should be shared by all invocations of this function. It should use a thread_local semaphore (exactly as in the "external semaphore" example in the tutorial) to limit the total number of probes happening in parallel, from any file.

**Comment by slivne** [comment]
#1275 should fix part of this  and make sure we search for files only on shard 0 and then propegate the file lists to other shards

**Comment by nyh** [comment]
Avi and Benoit proposed an orthogonal method of reducing the duplicate probing work on each shard (we probably need to implement both): We use the sstable's generation number modulo the number of shards to decide which shard will probe this sstable. Only this shard will probe the sstable, decide which shard needs to open it, and send a message to other shards which might need to open this.

**Comment by nyh** [comment]
Oh, yes, that orthogonal idea is #1275. I think we still need to limit the amount of parallelism, even on one shard.

**Comment by avikivity** [comment]
I'm surprised we don't have a parallel_for_each variant that limits concurrency for you.  People shouldn't have to do the tricks with the semaphore each time.

**Comment by slivne** [comment]
@raphaelsc did your work on loading sstables / resharding  (as part of moving to use other bits for partitioning data) solve this

**Comment by avikivity** [comment]
Not sure what's the problem in this bug.  Is parallel probing problematic?  It's only a problem if it consumes large amounts of memory.

**Comment by raphaelsc** [comment]
this issue can be closed after ef428d008cedb4861ddb69ca07c58e46c40a28ea
#2461 describes why unbounded parallelism was a problem.
